com.unity.shadergraph/Editor/Generation/Processors/CustomInterpolatorUtils.cs

```diff
@@ -54,16 +54,19 @@ internal struct Descriptor
             internal string name;     // for struct or function.
             internal string define;   // defined for client code to indicate we're live.
             internal string splice;   // splice location, prefer use something from the list.
+            internal string preprocessor;
 
             internal bool isBlock => src != null && dst != null && name == null && splice != null;
             internal bool isStruct => src == null && dst == null && name != null && splice != null;
             internal bool isFunc => src != null && dst != null && name != null && splice != null;
             internal bool isDefine => define != null && splice != null && src == null && dst == null & name == null;
+            internal bool isValid => isDefine || isBlock || isStruct || isFunc;
+            internal bool hasPreprocessor => preprocessor != null && preprocessor != "";
 
-            internal static Descriptor MakeFunc(string splice, string name, string dstType, string srcType, string define = "") => new Descriptor { splice = splice, name = name, dst = dstType, src = srcType, define = define };
-            internal static Descriptor MakeStruct(string splice, string name, string define = "") => new Descriptor { splice = splice, name = name, define = define };
-            internal static Descriptor MakeBlock(string splice, string dst, string src) => new Descriptor { splice = splice, dst = dst, src = src };
-            internal static Descriptor MakeDefine(string splice, string define) => new Descriptor { splice = splice, define = define };
+            internal static Descriptor MakeFunc(string splice, string name, string dstType, string srcType, string define = "", string preprocessor = "") => new Descriptor { splice = splice, name = name, dst = dstType, src = srcType, define = define, preprocessor = preprocessor };
```

Comment:
Is it a problem that all of these share the same defines?

Reply:
Shouldn't be a proble, what do you mean?

Reply:
I wasn't sure if it could be problem or not. Just wanted to ask you if it could be. Basically wasn't sure if different defines might be needed for blocks, structs, defines, or functions.

Reply:
Oh sure-- uh... yea, here's a bit more explanation than what you're probably looking for--

Each descriptor is pseudo-typed based on the static construction function used. A descriptor instance does not represent more than one of these types at a time. Even then, the define is used the same way in all cases-- it's appended at the end of the code generated for the descriptor so that subsequent template code can branch off of it (primarily for cases where nothing is generated).

Quick example would be to look at Varyings.hlsl in URP, you can see the preprocessor wrapped call for copying custom interpolators into the VertexDescription object:
```
#if defined(CUSTOMINTERPOLATOR_VARYPASSTHROUGH_FUNC)
    CustomInterpolatorPassThroughFunc(output, vertexDescription);
#endif
```

The define comes from UniversalTarget.cs, which describes how and where the above function gets generated. The define is used to allow the callsite to guard itself.
```
CustomInterpSubGen.Descriptor.MakeFunc(CustomInterpSubGen.Splice.k_splicePreSurface, "CustomInterpolatorPassThroughFunc", "Varyings", "VertexDescription", "CUSTOMINTERPOLATOR_VARYPASSTHROUGH_FUNC")
```

Reply:
Oh yeah...it's effectively a static constructor, missed that. Makes sense!